### PR TITLE
CHANGE (GraphEngine): @W-12672065@: Handles backend changes to cover remaining NPE Rule cases.

### PR DESCRIPTION
**Pull request by rmohan20**
1. Added fix to null check priorities on indeterminant values
2. Added changes to throw null access checks when evaluating binary expressions
3. Minor changes to ApexNullPointerExceptionRule to accommodate Binary expression vertices.
4. Test changes
5. Modified user-facing text to handle method calls and binary expressions.

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/config/UserFacingMessages.java [hunk 1]
```diff
@@ -16,7 +16,7 @@ public static final class RuleDescriptions {
 
     public static final class RuleViolationTemplates {
         public static final String APEX_NULL_POINTER_EXCEPTION_RULE =
-                "Operation [%s] is likely to throw a NullPointerException";
+                "%s is likely to throw a NullPointerException";
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/config/UserFacingMessages.java:19]: The display string cannot get the lhs and rhs values on violation message without many more low level changes. Based on priorities, I think modifying the text to fit the context is easier.

File: sfge/src/main/java/com/salesforce/graph/ops/ApexValueUtil.java [hunk 2]
```diff
@@ -33,15 +34,15 @@
 
 public final class ApexValueUtil {
     private static final Logger LOGGER = LogManager.getLogger(ApexValueUtil.class);
+    private static final String UNKNOWN = "UNKNOWN";
 
     /** Currently concatenates string values. TODO: Expand to other operations on Integers etc. */
     public static Optional<ApexValue<?>> applyBinaryExpression(
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/ApexValueUtil.java:40]: Change in order and minor code refactor to perform null access checks on binary vertices.

File: sfge/src/main/java/com/salesforce/graph/ops/ApexValueUtil.java [hunk 3]
```diff
@@ -51,28 +52,57 @@ public static Optional<ApexValue<?>> applyBinaryExpression(
...
+
+    private static Optional<ApexValue<?>> performStringConcatenation(
+            ApexStringValue lhs,
+            ApexStringValue rhs,
+            BinaryExpressionVertex vertex,
+            SymbolProvider symbols) {
+        final String lhsValue =
+                lhs.isValuePresent() ? lhs.getValue().get() : lhs.isNull() ? "null" : UNKNOWN;
+        final String rhsValue =
+                rhs.isValuePresent() ? rhs.getValue().get() : rhs.isNull() ? "null" : UNKNOWN;
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/ApexValueUtil.java:92]: we were previously ignoring this null altogether. Apex, however, treats null strings as "null". I've made relevant changes here.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexValue.java [hunk 4]
```diff
@@ -220,16 +220,22 @@ public boolean isNotNull() {
      * null or it is a an uninitialized variable which will be initialized to null by the system.
      */
     public boolean isNull() {
-        if (ValueStatus.UNINITIALIZED.equals(status)) {
-            // Uninitialized values are initialized to null by the compiler
-            return true;
-        } else if (ValueStatus.INDETERMINANT.equals(status)) {
-            return false;
+        boolean returnValue =
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexValue.java:223]: This was the bug that didn't count null constraints in indeterminant strings. Fixed the order of checks.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ObjectPropertiesHolder.java [hunk 5]
```diff
@@ -167,7 +167,7 @@ public ApexValue<?> put(ApexValue<?> key, ApexValue<?> value) {
              * <p>SObject sObj = new Account(); String keyName = null; for (some iteration) { if
              * (some comparison) { keyName = 'field1'; } } sObj.put(keyName, 'value');
              */
-            throw new NullValueAccessedException(key, null);
+            throw new NullValueAccessedException(key, (MethodCallExpressionVertex) null);
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ObjectPropertiesHolder.java:170]: Not needed anymore. Will revert.

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**